### PR TITLE
Add CUTLASS-based support for mixed dtypes matrix multiplication

### PR DESCRIPTION
**Pull request by alexsamardzic**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #109712
* __->__ #107782

cc @ptrblck

### CODE REVIEW

**Comment by alexsamardzic** [comment]
This PR is mainly to support multiplying a matrix with another matrix that comes from quantization.  It is based on work from [FasterTransformer](https://github.com/NVIDIA/FasterTransformer) project.

The `cutlass_extensions` directory is a copy of `src/fastertransformer/cutlass_extensions/include/cutlass_extensions` directory from FasterTransformer project, with a minor change (inlining the contents of `src/fastertransformer/utils/cuda_bf16_wrapper.h` instead of having it included).  It contains CUTLASS extensions needed to apply scale factor to quantized values.

The FasterTransformer project uses CUTLASS 2.10, thus in order to build this PR, one has to replace `third_party/cutlass` in the PyTorch source tree with CUTLASS 2.10 source code.  (But then, one must not rebase after [PR 107782](https://github.com/pytorch/pytorch/pull/107782) merged, as this one won't work without CUTLASS 3.x).

**Comment by alexsamardzic** [comment]
A script to compare performance between doing linear operator on original matrices (based on cuBLAS), and using the first matrix unchanged and the second one quantized (based on this PR, and then another version based on code compiled to Triton):

<details>
 <summary>Benchmarking script</summary>

```Python
import random
import torch
import torch.utils.benchmark as benchmark

from itertools import product
from torch.testing import make_tensor

import torch._inductor.config as inductor_config

device = 'cuda'
dtype = torch.float16
dtypeq = torch.uint8

torch.set_printoptions(precision=3, threshold=None, edgeitems=4, linewidth=460, profile=None, sci_mode=False)

@inductor_config.patch(force_mixed_mm=True, max_autotune_gemm=True)
def mm_triton(a, b, c, d):
    return torch.mm(a, b.to(a.dtype)) * c + d

if __name__ == '__main__':
    results = []

    shapes = [
        #(8, 64, 32),
        # distilbert shapes
        (768, 3072, 768),
        (3072, 768, 3072),
        # jiecao shapes
        (1024, 1536, 2048),
        (1024, 9408, 2048),
        (1024, 3200, 2048),
        (1024, 256, 9472),
        (1024, 10240, 256),
        (1024, 256, 12608),
        (1024, 2560, 1024),
        (1024, 512, 10240),
        (1024, 10240, 512),
        (1024, 2048, 1024),
        (1024, 512, 512),
        (1024, 1024, 1024),
        (1024, 2048, 2048),
        (2048, 1536, 2048),
        (2048, 9408, 2048),
        (2048, 3200, 2048),
        (2048, 256, 9472),
        (2048, 10240, 256),
        (2048, 256, 12608),
        (2048, 2560, 1024),
        (2048, 512, 10240),
        (2048, 10240, 512),
        (2048, 2048, 1024),
        (2048, 512, 512),
        (2048, 1024, 1024),
        (2048, 2048, 2048),
    ]
    batch_sizes = [1] ###[4, 16, 64, 256]


    for (m, k, n), batch_size in product(shapes, batch_sizes):
        print(m, k, n, batch_size)
        # label and sub_label are the rows
        # description is the column
        ###try:
        if True:
            label = 'cuBLAS vs. CUTLASS vs. Triton mixed dtypes GEMM'
            sub_label = f'm:{m:5d} | k:{k:5d} | n:{n:5d} | batch_size: {batch_size:4d}'

            fp16_low, fp16_high = 0.5, 1.5
            ui8_low, ui8_high = 129, 130 #FIXME: make this range broader
            a = make_tensor(m, k, low=fp16_low, high=fp16_high, dtype=torch.float16, device=device)
            b = make_tensor(k, n, low=ui8_low, high=ui8_high, dtype=torch.uint8, device=device)
            scale = make_tensor((1, n), low=fp16_low, high=fp16_high, dtype=a.dtype, device=device)
            bias = make_tensor((1, n), low=fp16_low, high=fp16_high, dtype=a.dtype, device=device)

            b_ref = b - 128
            weights_ref = b_ref.to(a.dtype) * scale.expand((k, n))
            
            c_ref = torch.mm(a, weights_ref) + bias

            c = torch.ops.aten._fp16_uint8_mm(a, b, scale, bias)
            torch.testing.assert_close(c, c_ref, rtol=1e-3, atol=0)

            mm_triton_compiled = torch.compile(mm_triton)            
            c = mm_triton_compiled(a, b_ref, scale, bias)
            torch.testing.assert_close(c, c_ref, rtol=1e-2, atol=0)
            
            measurement = benchmark.Timer(
                stmt='mm(a, weights) + bias',
                globals={'mm': torch.mm, 'a': a, 'weights': weights_ref, 'bias': bias},
                label=label,
                sub_label=sub_label,
                description='cuBLAS latency',
            ).blocked_autorange()
            results.append(measurement)

            measurement = benchmark.Timer(
                stmt='mm(a, b, scale, bias)',
                globals={'mm': torch.ops.aten._fp16_uint8_mm, 'a': a, 'b': b, 'scale': scale, 'bias': bias},
                label=label,
                sub_label=sub_label,
                description='CUTLASS latency',
            ).blocked_autorange()
            results.append(measurement)

            measurement = benchmark.Timer(
                stmt='mm(a, b, scale, bias)',
                globals={'mm': mm_triton_compiled, 'a': a, 'b': b_ref, 'scale': scale, 'bias': bias},
                label=label,
                sub_label=sub_label,
                description='Triton latency',
            ).blocked_autorange()
            results.append(measurement)

        ###except Exception:
            continue

        # with profile(activities=[ProfilerActivity.CUDA, ProfilerActivity.CPU], record_shapes=True, with_stack=True) as prof:
        #     with record_function('CUTLASS'):
        #         sparse_linear(input_tensor)

        # prof.export_stacks(f'{m}_{k}_{n}_cutlass_profiler_stacks.txt', 'self_cuda_time_total')

    compare = benchmark.Compare(results)
    compare.colorize(rowwise=True)
    compare.print()
```
</details>

**Comment by alexsamardzic** [comment]
@cpuhrsch @drisspg @amjames

File: test/test_mixed_dtypes_mm.py
Comment (cpuhrsch) [inline test/test_mixed_dtypes_mm.py:4]: You can put this into test_linalg.py

Comment (alexsamardzic) [reply test/test_mixed_dtypes_mm.py]: Done.

File: test/test_linalg.py
Comment (IvanYashchuk) [inline test/test_linalg.py:1]: This file is dedicated to testing `torch.linalg` namespace and not matrix multiplications.

**Review by IvanYashchuk — CHANGES_REQUESTED** [review]

File: test/test_mixed_dtypes_mm.py
Comment (alexsamardzic) [reply test/test_mixed_dtypes_mm.py]: Done, thanks!

File: test/test_linalg.py
Comment (cpuhrsch) [reply test/test_linalg.py:1]: It has tests for int8 matmul, why not for int8 X fp16 matmul?

Comment (cpuhrsch) [reply test/test_linalg.py:1]: But in any case, test_matmul_cuda.py seems like the right spot.

**Comment by alexsamardzic** [comment]
Just a note to keep an eye [here](https://github.com/NVIDIA/cutlass/discussions/911), for having mixed dtypes GEMM support upstream in CUTLASS.

File: aten/src/ATen/native/cuda/MixedDtypesMM.cu
Comment (cpuhrsch) [inline aten/src/ATen/native/cuda/MixedDtypesMM.cu:154]: Can we make bias optional? That means generating a specialization with bias fusion enabled and a specialization without bias fusion.

**Comment by alexsamardzic** [comment]
This PR is ready for review, `_mixed_dtypes_linear()` linear operator is now implemented for float16/bfloat16 and quantized int8 arguments, together with optional bias argument and support for optionally fusing ReLU and SiLU activations.

Weight matrix for linear operator has to be reordered, according to the CUTLASS requirements, up-front.  This is implemented by `quantized_weight_reorder_for_mixed_dtypes_linear()` function in `torch/quantization/_quantized_conversions.py`.  The output of this function could be then passed any number of times as weight argument to `_mixed_dtypes_linear()` operator.

The CUTLASS extensions re-used from FasterTransformer projects are put in `cutlass_extensions` directory, in the PyTorch top-level directory (any suggestion for moving them elsewhere is welcome); anything in-there should be skipped during a review.  Since these comprise large chunks of code, this PR is above LOC limit, and will need additional approval for eventual merge.  Also, this code is not yet processed through linter - as usual, the catch is that code in the other project (FasterTransformer in this case) is not linted, but on the other side if linted it will be hard to remember changes made in this code for use in PyTorch (in this case, porting it from CUTLASS 2.10 to CUTLASS 3.x).

**Comment by cpuhrsch** [comment]
Great news @alexsamardzic !

Could you move cutlass_extensions under aten/src/ATen/native/cuda?

**Comment by alexsamardzic** [comment]
> Could you move cutlass_extensions under aten/src/ATen/native/cuda?

Done.

**Comment by alexsamardzic** [comment]
A script to compare performance between doing linear operator on original matrices (based on cuBLAS), and using the first matrix unchanged and the second one quantized (based on this PR, and then another version based on code compiled to Triton):

<details>
 <summary>Benchmarking script</summary>

```Python
import random
import torch
import torch.utils.benchmark as benchmark

from itertools import product
from torch.testing import make_tensor
from torch.quantization._quantized_conversions import (
    quantized_weight_reorder_for_mixed_dtypes_linear,
)

import torch._inductor.config as inductor_config

device = 'cuda'
dtype = torch.float16
dtypeq = torch.int8
dtypeqq = torch.uint8

torch.set_printoptions(precision=3, threshold=None, edgeitems=4, linewidth=460, profile=None, sci_mode=False)

@inductor_config.patch(force_mixed_mm=True, max_autotune_gemm=True)
def mixed_dtypes_linear_triton(input, weight, scale, bias):
    return torch.mm(input, weight.to(input.dtype)) * scale + bias

if __name__ == '__main__':
    results = []

    shapes = [
        #(8, 64, 32),
        # distilbert shapes
        (768, 3072, 768),
        (3072, 768, 3072),
        # jiecao shapes
        (1024, 1536, 2048),
        (1024, 9408, 2048),
        (1024, 3200, 2048),
        (1024, 256, 9472),
        (1024, 10240, 256),
        (1024, 256, 12608),
        (1024, 2560, 1024),
        (1024, 512, 10240),
        (1024, 10240, 512),
        (1024, 2048, 1024),
        (1024, 512, 512),
        (1024, 1024, 1024),
        (1024, 2048, 2048),
        (2048, 1536, 2048),
        (2048, 9408, 2048),
        (2048, 3200, 2048),
        (2048, 256, 9472),
        (2048, 10240, 256),
        (2048, 256, 12608),
        (2048, 2560, 1024),
        (2048, 512, 10240),
        (2048, 10240, 512),
        (2048, 2048, 1024),
        (2048, 512, 512),
        (2048, 1024, 1024),
        (2048, 2048, 2048),
    ]
    batch_sizes = [1, 4]


    for (m, k, n), batch_size in product(shapes, batch_sizes):
        print(m, k, n, batch_size)
        # label and sub_label are the rows
        # description is the column
        try:
            label = 'cuBLAS vs. CUTLASS vs. Triton mixed dtypes GEMM'
            sub_label = f'm:{m:5d} | k:{k:5d} | n:{n:5d} | batch_size: {batch_size:4d}'

            val_lo, val_hi = -1, 1
            valq_lo, valq_hi = -2, 2
            input = make_tensor(batch_size, m, k, low=val_lo, high=val_hi, dtype=torch.float16, device=device)
            weight = make_tensor(n, k, low=valq_lo, high=valq_hi, dtype=dtypeq, device=device)
            scale = make_tensor((n,), low=val_lo, high=val_hi, dtype=input.dtype, device=device)
            bias = make_tensor((n,), low=val_lo, high=val_hi, dtype=input.dtype, device=device)

            input_ref = input.reshape(-1, input.shape[-1])
            weight_ref = weight.to(input.dtype) * scale.view(n, 1)
            bias_ref = bias.view(1, n)
            output_ref = torch.nn.functional.linear(
                input_ref, weight_ref, bias=bias_ref
            ).reshape(*input.shape[:-1], n)

            weight_cutlass = quantized_weight_reorder_for_mixed_dtypes_linear(weight).view(
                    dtypeqq
            )
            output = torch.ops.aten._mixed_dtypes_linear(
                input,
                weight_cutlass,
                scale,
                bias=bias
            )
            ###torch.testing.assert_close(output, output_ref, rtol=1e-3, atol=0)

            mixed_dtypes_linear_triton_compiled = torch.compile(mixed_dtypes_linear_triton)
            weight_triton = weight.T.contiguous()
            output = mixed_dtypes_linear_triton_compiled(input_ref, weight_triton, scale, bias).reshape(*input.shape[:-1], n)
            ###torch.testing.assert_close(output, output_ref, rtol=1e-2, atol=0)
            
            measurement = benchmark.Timer(
                stmt='linear(input, weight, bias)',
                globals={'linear': torch.nn.functional.linear, 'input': input_ref, 'weight': weight_ref, 'bias': bias_ref},
                label=label,
                sub_label=sub_label,
                description='cuBLAS latency',
            ).blocked_autorange()
            results.append(measurement)

            measurement = benchmark.Timer(
                stmt='linear(input, weight, scale, bias=bias)',
                globals={'linear': torch.ops.aten._mixed_dtypes_linear, 'input': input, 'weight': weight_cutlass, 'scale': scale, 'bias': bias},
                label=label,
                sub_label=sub_label,
                description='CUTLASS latency',
            ).blocked_autorange()
            results.append(measurement)

            measurement = benchmark.Timer(
                stmt='linear(input, weight, scale, bias)',
                globals={'linear': mixed_dtypes_linear_triton_compiled, 'input': input_ref, 'weight': weight_triton, 'scale': scale, 'bias': bias},
                label=label,
                sub_label=sub_label,
                description='Triton latency',
            ).blocked_autorange()
            results.append(measurement)

        except Exception:
            continue

    compare = benchmark.Compare(results)
    compare.colorize(rowwise=True)
    compare.print()
```
</details>

**Review by cpuhrsch — APPROVED** [review]
Accepting this so I can check internal build configuration.

Don't merge yet @alexsamardzic

File: aten/src/ATen/native/cuda/cutlass_extensions/epilogue_helpers.h
Comment (cpuhrsch) [inline aten/src/ATen/native/cuda/cutlass_extensions/epilogue_helpers.h:12]: Unfortunately all of these includes will need to use "<" and ">" instead of quotes.

Comment (alexsamardzic) [reply aten/src/ATen/native/cuda/cutlass_extensions/epilogue_helpers.h:12]: No problem, these files have to go through the linter anyway, it will fix it.  I was just intending to do it just before the merge, so that I could save their current versions at that point, namely I had to made some changes vs. originals copied from FasterTransfomer (for update to CUTLASS 3.x, etc.), but once they linted according to the PyTorch lint rules, there will be many more differences.

Comment (cpuhrsch) [reply aten/src/ATen/native/cuda/cutlass_extensions/epilogue_helpers.h:12]: Yeah since we're including these files directly I don't think we can dodge these constraints.

**Comment by cpuhrsch** [comment]
@alexsamardzic - Alright, let's get into landing this. Could you try resubmitting this as a standalone PR (without ghstack) and on top of the latest main commit? Thank you.

**Comment by alexsamardzic** [comment]
Merged as #110981.

**Pull request closed without merging**
